### ISSUE TITLE
Factory Unification

### ISSUE THREAD

**Issue by ef4** [samselikoff/ember-cli-mirage#940]
# Motivation

Mirage's `Model` is hard to teach because it falls into an uncanny valley of seeming like the same thing as `DS.Model`. It doesn't do much other than declare that a type exists and that it has certain associations.

# Proposal

`Factory` makes a better foundational primitive for Mirage. There's no reason we can't annotate relationships on factories instead of on models, and drop the concept of Model.

# On the original motivation for Model

https://github.com/samselikoff/ember-cli-mirage/issues/69 lays out the original thinking. I think all the objections against declaring associations within factories can be overcome.

> ...but then, users would need to ensure all their relationship metadata is in their factories. What if this doesn't match up with their business requirements regarding what makes their models valid? For example, what if a contact didn't need an address to be valid, and so the factory didn't have the association in its definition?

This is a circular argument about the API of Factory. There's no reason we can't allow Factory to declare associations that don't have default implementations. For example, 

```js
address: hasOne()
```

can declare an association without providing any default implementation, whereas

```js
address: hasOne(function(n, server) {
  return server.create('address');
})
```

can both declare the association and provide a default value. This also eliminates the need to imperatively create related records in `afterCreate`.

# On Foreign Keys

Several discussions about associations have unnecessarily entangled the question of foreign keys.

Foreign keys _are an implementation detail_. Consider that JSON:API _deliberately hides them within the server's implementation_. A spec-compliant JSON:API server is expected to be able to work with no publicly-visible foreign keys at all.

Mirage has the luxury of being an in-memory data store, so indirection via keys is not necessary. Relationships can point directly at the objects themselves.

When users want to expose a property that acts as a foreign key to a relationship, that can be done directly:

```js
  get addressId() {
    if (this.address) {
      return this.address.id;
    }  
  }
  set addressId(id) {
    this.address = server.addresses.find(id);
  }
```

This pattern can readily be packaged up as a macro since it is a useful common case:

```js
  addressId: foreignKeyFor('address', /* { remoteKey: 'id' } */),
```

# On Fixtures

Mirage supports fixtures as a parallel world apart from using factories at all. One may object to making factories the foundational primitive since they seem extraneous when using fixtures.

But I think factories should actually run anyway when using fixtures. The contents of the fixture can be treated as inputs to the factory instead of being raw, fully-specified records. This is how I actually use fixtures in practice, because otherwise the fixture files are forced to be more fragile and verbose.

For example, `server.loadFixtures('articles')` is conceptually shorthand for

```js
import articlesFixtures from 'mirage/fixtures/articles';
articlesFixtures.forEach(a => server.create('article', a));
```

**Comment by samselikoff** [comment]
Thanks for taking the time to write this up and share your thoughts!

This comes at a good time as I am reworking parts of the ORM, and was going to publish some small breaking changes in 0.3. It could work out that we abstract away `Model` from Mirage's public interface either before 0.3 or v1.0.

---

Conceptually I liked the pattern of factory properties being "things you get when creating an object from this factory." Factories like this seem like they could be hard to parse

```js
// factories/post.js

Mirage.Factory.extend({

  user: hasOne(function(n, server) {
    return server.create('address');
  }),

  comments: hasMany(),

  title: 'Foo',

  publishedAt: '2016-01-01'
  
})
```

(in this case, `server.create('post')` would give you an associated user but no comments), but I could be hand-wringing over nothing.

There is also an open PR for an [association helper](https://github.com/samselikoff/ember-cli-mirage/pull/880) that I expect we'll merge soon, which would address the point about needing `afterCreate` to make & wire up standard associations.

I wonder if we could have a "declaration" section for factories, as a clear place where relationship metadata is encoded:

```js
Mirage.Factory.extend({

  _schema: {
    user: hasOne(),
    comments: hasMany()
  }

})
```

Is there a decorators solution here?

```js
Mirage.Factory.extend({

  @hasOne user
  @hasMany comments

  user: association(),
  
  title: 'Foo',

  publishedAt: '2016-01-01'
})
```

Ultimately, I agree that it might not be worth keeping both concepts anymore. The original motivation was to have one place where your schema is defined (rails AR, schema.rb) and one place where fake data is managed/produced (factory_girl). It's clearly caused lots of confusion so it's worth discussing.

Another point, if we dropped Models from the public interface, we'd still need `schema.users.create()`. Would it be more confusing to explain what this is doing without ever having introduced the concept of the model? Would we add an API to factories/server.create to "make an empty object of type X"?

**Comment by samselikoff** [comment]
Also, once the ORM supports all the relationship types ED does, I want Mirage to autoread ED's schema. This could alleviate some of the confusion.

**Comment by ef4** [comment]
I like the decorators idea. It's also possible to use named exports alongside the factory class:

```js
export const schema = {
  user: hasOne(),
  comments: hasMany()
};
export default Factory.extend({...});
```

**Comment by samselikoff** [comment]
Re-reading this now here are my thoughts:

 - Having Mirage automatically read Ember Data models and generate its own models at runtime should address a lot of the confusion
 - Requiring factories for each model would be a breaking API change
 - We may eventually add default values to Models (different from factory values), if these were on the same object it could be more confusing
 - The roles for models and factories are borrowed from backend systems, which does provide some point of comparison for folks familiar with those systems. Also, we don't have to invent our own abstractions, we're borrowing well-worn patterns

For these reasons I think I'm going to close this for now. I like your fixtures idea, I am open to adding that to the API. The foreign keys in 0-3 ended up being different from your proposal, I had some issues with an in-memory graph but can't recall them now.

If you (or anyone) has more thoughts here that will turn into actionable items, I'm more than happy to reopen this so we can discuss. Thanks again!
